src/function/midi.rs

```diff
@@ -0,0 +1,118 @@
+//! Musical Instrument Digital Interface (MIDI) function.
+//!
+//! The Linux kernel configuration option `CONFIG_USB_CONFIGFS_F_MIDI` must be enabled. Can use `amidi -l` once the gadget is configured to list the MIDI devices.
+//!
+//! # Example
+//!
+//! ```no_run
+//! use usb_gadget::function::midi::Midi;
+//! use usb_gadget::{default_udc, Class, Config, Gadget, Id, Strings};
+//!
+//! let mut builder = Midi::builder();
+//! // This must be an available sound device index - see docs on index for more information
+//! builder.index = 0;
+//! builder.id = "midi".to_string();
+//! builder.in_ports = 1;
+//! builder.out_ports = 1;
+//! let (midi, func) = builder.build();
+//!
+//! let udc = default_udc().expect("cannot get UDC");
+//! let reg =
+//!     // USB device descriptor base class 0, 0, 0: use Interface Descriptors
+//!     // Linux Foundation VID Gadget PID
+//!     Gadget::new(Class::new(0, 0, 0), Id::new(0x1d6b, 0x0104), Strings::new("Clippy Manufacturer", "Rust MIDI", "RUST0123456"))
+//!         .with_config(Config::new("MIDI Config 1").with_function(func))
+//!         .bind(&udc)
+//!         .expect("cannot bind to UDC");
+//!
+//! println!(
+//!     "USB MIDI {} at {} to {} status {:?}",
+//!     reg.name().to_string_lossy(),
+//!     reg.path().display(),
+//!     udc.name().to_string_lossy(),
+//!     midi.status()
+//! );
+//! ```
+
+
+use std::{ffi::OsString, io::Result};
+
+use super::{
+    util::{FunctionDir, Status},
+    Function, Handle,
+};
+
+/// Builder for USB musical instrument digital interface (MIDI) function.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct MidiBuilder {
+    /// MIDI buffer length
+    pub buflen: u16,
+    /// ID string for the USB MIDI adapter
+    pub id: String,
+    /// Number of MIDI input ports
+    pub in_ports: u8,
+    /// Number of MIDI output ports
+    pub out_ports: u8,
+    /// Sound device index for the MIDI adapter
+    ///
+    /// There must be a sound device available with this index. If the device fails to register and in dmesg one sees 'cannot find the slot for index $index (range 0-1), error: -16', then the index is not available. Most likely the index is already in use by the physical sound card. Try another index within range or unload the physical sound card driver. See [USB Gadget MIDI](https://linux-sunxi.org/USB_Gadget/MIDI).
+    pub index: u8,
```

Comment:
If `index` is unset, the kernel should automatically select an available sound device index.

https://github.com/torvalds/linux/blob/7af08b57bcb9ebf78675c50069c54125c0a8b795/drivers/usb/gadget/function/f_midi.c#L1272